### PR TITLE
External/Global services

### PR DESCRIPTION
**Pull request by alena1108**

### CODE REVIEW

**Comment by alena1108** [comment]
@ibuildthecloud Darren, could you please review. After this PR is merged, I will have to change Healtcheck branch to exclude External serviceExpose targets from the healtcheck

**Comment by alena1108** [comment]
Will include Global services implementation to this PR as well

File: code/iaas/config-item/server/src/main/java/io/cattle/platform/configitem/context/dao/impl/DnsInfoDaoImpl.java [hunk 1]
```diff
@@ -249,4 +250,65 @@ protected DnsEntryData map(List<Object> input) {
...
+                .on(clientNicIpTable.NIC_ID.eq(clientNic.ID))
+                .join(clientIpAddress)
+                .on(clientNicIpTable.IP_ADDRESS_ID.eq(clientIpAddress.ID))
+                .join(clientInstance)
+                .on(clientNic.INSTANCE_ID.eq(clientInstance.ID))
+                .where(NIC.INSTANCE_ID.eq(instance.getId())
+                        .and(NIC.VNET_ID.isNotNull())
+                        .and(NIC.REMOVED.isNull())
+                        .and(clientIpAddress.ROLE.eq(IpAddressConstants.ROLE_PRIMARY))
+                        .and(clientIpAddress.REMOVED.isNull())
```
Comment (ibuildthecloud) [inline code/iaas/config-item/server/src/main/java/io/cattle/platform/configitem/context/dao/impl/DnsInfoDaoImpl.java]: Do we need a clientIpAddress.ADDRESS.notNull()?

Comment (alena1108) [inline code/iaas/config-item/server/src/main/java/io/cattle/platform/configitem/context/dao/impl/DnsInfoDaoImpl.java]: I've added a validtion for that to the *ftl file itself, but if you think its better to place it here - I can do that

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java [hunk 2]
```diff
@@ -284,40 +291,41 @@ private void populateExternalLinksForService(Service service, Map<String, Object
...
+                    }
                 }
+
+                launchConfigItems.put(key, dataObj);
             }
-            launchConfigItems.put(ServiceDiscoveryConfigItem.VOLUMESFROM.getCattleName(), volumesFrom);
         }
 
+        // 3. add extra parameters
+        Object registryCredentialId = serviceData.get(ServiceDiscoveryConfigItem.REGISTRYCREDENTIALID
```
Comment (ibuildthecloud) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java]: Why do we need this?

File: resources/content/db/changelog.xml [hunk 3]
```diff
@@ -45,4 +45,5 @@
     <include file="db/core-038.xml"/>
     <include file="db/core-039.xml"/>
     <include file="db/core-040.xml"/>
+    <include file="db/external.xml"/>
```
Comment (ibuildthecloud) [inline resources/content/db/changelog.xml]: Rename before merge

**Comment by alena1108** [comment]
@ibuildthecloud Darren, unrelated test - test_docker - failed, but otherwise the PR is squashed and ready to be merged.

**Pull request merged**